### ISSUE TITLE
"rewrite in *" fails where "rewrite in H1" works

### ISSUE THREAD

**Issue by Zimmi48** [Zimmi48/bugzilla-test#3051]
Note: the issue was created automatically with bugzilla2github tool

# Bugzilla Bug ID: 3051
Date: 2013-05-25 18:56:12 +0200
From: ruibaptista@live.com
To:   
CC:   coq-bugs-redist@lists.gforge.inria.fr, @_JasonGross, @_ppedrot

Last updated: 2014-01-19 04:01:57 +0100

## Bugzilla Comment ID: 5907
Date: 2013-05-25 18:56:12 +0200
From: ruibaptista@live.com

In the code bellow "rewrite (A1 s1 s2 e1 e2 _ H1) in *." fails where "rewrite (A1 s1 s2 e1 e2 _ H1) in H2." doesn't.

Inductive Prod (s1 s2 : Set) : Set :=
  | Pair : s1 -> s2 -> Prod s1 s2.

Arguments Pair {_ _} _ _.

Axiom A1 : forall (s1 s2 : Set) (e1 : s1) (e2 : s2) (p1 : Prod s1 s2),
  True -> p1 = Pair e1 e2.

Lemma L1 : forall (s1 s2 : Set) (e1 e3 : s1) (e2 e4 : s2)
  (p1 : Prod s1 s2) (p2 : Prod s1 s2), True -> p1 = Pair e1 e2 ->
  p2 = Pair e3 e4 -> False.
Proof.
intros s1 s2 e1 e3 e2 e4 p1 p2 H1 H2 H3.
rewrite (A1 s1 s2 e1 e2 _ H1) in *.
rewrite (A1 s1 s2 e1 e2 _ H1) in H2.
rewrite (A1 s1 s2 e1 e2 _ H1) in H3.
Qed.

## Bugzilla Comment ID: 6442
Date: 2014-01-19 01:55:39 +0100
From: @_ppedrot

This seems intentional from the comments of the code implementing this tactic. There is a comment in tactics/equality.ml stating:

(* If the term to rewrite uses an hypothesis H, don't rewrite in H *)

I do not know if this is relevant, but this is precisely what forbids the rewriting in H2 when using *, because the underscore in (A1 s1 s2 e1 e2 _ H1) produces an evar depending on H2. Any comment about removing this check anyone?

## Bugzilla Comment ID: 6443
Date: 2014-01-19 03:59:29 +0100
From: @_JasonGross

I think that the check should not be bypassed, and instead that any hypotheses that fail the check should not be rewritten in.  To me, "in *" means "in all valid hypothesis" not "in all hypotheses", so I expect it to succeed unless it would fail on every single hypothesis if I did it manually.

## Bugzilla Comment ID: 6444
Date: 2014-01-19 04:01:57 +0100
From: @_JasonGross

Oh, sorry I misread, missed "evar depending on H2", and thought that this was a problem that I had had in the past where "rewrite foo in *" failed but "rewrite foo in H" didn't (no evars).  (Is it the same issue?)  Anyway, can the check be deferred until all arguments are filled in, or can we use a fresh evar for each hypothesis such that that evar doesn't depend on the hypothesis being rewritten into?
